capstone-app/src/app/modules/history/wiki-results.service.ts

```diff
@@ -17,15 +19,74 @@ export class WikiResultsService {
     'https://en.wikipedia.org/w/api.php?origin=*&action=parse&page=';
   private static readonly URL_END = '&format=json';
 
-  search(query: string): Observable<WikiSearchResult> {
+  search(query: string): Observable<{title: string, history: string, furtherReading: Map<string, string>}> {
     return this.searchHandler
       .getQueryString(query)
       .pipe(
         switchMap((result) =>
           this.http.get<WikiSearchResult>(
             `${WikiResultsService.URL_BEGINNING}${result}${WikiResultsService.URL_END}`
           )
-        )
+        ),map(history => {
+          if (history.parse.text) {
+            const historyResult = this.fixString(history.parse.text['*']);
+            const title = history.parse.title;
+            return {title, ...historyResult};
+          }
+          else {
+            throw new Error("API did not return a valid response");
+          }
+        }),
+        catchError(() => {
+          console.log('API did not return a valid response.');
+          const title = '';
+          const history = '';
+          const furtherReading = new Map<string, string>();
+          return of({title, history, furtherReading});
+        })
       );
   }
+
+  fixString(text: string): {history: string, furtherReading: Map<string, string>} {
+    const firstIndex = text.indexOf('<span class="mw-headline" id="History">History</span>', 0);
+    if (firstIndex !== -1){
+      const firstPartOfString = text.substring(firstIndex, text.length);
+      const endIndex = firstPartOfString.indexOf('<h2>', 0);
+      const startIndex = firstPartOfString.indexOf('</h2>', 0);
+      let middleOfString = firstPartOfString.substring(startIndex, endIndex);
+      const paragraphs = middleOfString.split('<p>');
+      if (paragraphs.length > 6){
+        middleOfString = paragraphs.slice(0,5).join('');
+      }
+      const furtherReading = this.findHrefs(middleOfString);
+      let history = middleOfString.split(/<.*?>/g).join('');
+      history = history.split(/&.*?;/g).join('');
+      return {history, furtherReading};
+    }
+    console.log('The city page was found, but unfortunately there was no history paragraph found!');
+    let history = text.split(/<.*?>/g).join('');
+    history = history.split(/&.*?;/g).join('');
+    return {history, furtherReading: new Map()};
+  }
+
+  findHrefs(text: string): Map<string, string> {
+    const urls = new Map<string, string>();
+    const el = document.createElement('p');
+    el.innerHTML = text;
+    const hrefs = Array.from(el.querySelectorAll('a'));
+    let count = 1;
+    for (const h of hrefs) {
+      if (h.getAttribute('href').charAt(0) === '#' || (h.getAttribute('title') && h.getAttribute('title').startsWith('Edit'))){
+        delete hrefs[hrefs.indexOf(h)];
```

Comment:
I'm concerned about deleting elements of an array as we're iterating over them. Why do we need to remove the elements?

Reply:
I'm deleting them because if they start with a # or 'Edit' it's not a url to a page with more information for further reading. It's something like an wiki edit page, which we don't want to display to the user.

Reply:
Right, but the "hrefs" array is not used after this, this if block is just to prevent the link from going into the "urls" map. My concern is that deleting elements of an array as we're simultaneously iterating over it could cause urls to be skipped or other strange bugs, and I don't think we actually need to delete these elements from href, they just just need to be skipped. This line could instead just be a comment that says "// Do nothing for edit urls" or something like that.

Reply:
Ahh, true